src/client/method-unit-actions.c

```diff
@@ -308,6 +308,127 @@ static int method_thaw_unit_on(Client *client, char *node_name, char *unit) {
+                        return r;
+                }
+                units_count++;
+        }
+        return r;
+}
+
+static int method_reset_failed_on_node(UNUSED Client *client, UNUSED char *node_name) {
+
+        _cleanup_sd_bus_error_ sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_sd_bus_message_ sd_bus_message *result = NULL;
+        int r = 0;
+
+        r = assemble_object_path_string(NODE_OBJECT_PATH_PREFIX, node_name, &client->object_path);
+        if (r < 0) {
+                return r;
+        }
+
+        r = sd_bus_call_method(
+                        client->api_bus,
+                        BC_INTERFACE_BASE_NAME,
+                        client->object_path,
+                        NODE_INTERFACE,
+                        "ResetFailed",
+                        &error,
+                        &result,
+                        "");
+        if (r < 0) {
+                fprintf(stderr, "Failed to issue method call: %s\n", error.message);
+                return r;
+        }
+
+        return r;
+}
+
+static int method_reset_failed_on_all_nodes(UNUSED Client *client) {
+
+        int r = 0;
+        sd_bus *bus = NULL;
+
+        r = sd_bus_open_system(&bus);
+        if (r < 0) {
+                fprintf(stderr, "Failed to connect to system bus: %s", strerror(-r));
+                return r;
+        }
+
+        sd_bus_error error = SD_BUS_ERROR_NULL;
+        sd_bus_message *result = NULL;
+        r = sd_bus_call_method(
+                        bus,
+                        "org.eclipse.bluechi",
+                        "/org/eclipse/bluechi",
+                        "org.eclipse.bluechi.Controller",
+                        "ListNodes",
+                        &error,
+                        &result,
+                        "",
+                        NULL);
+        if (r < 0) {
+                fprintf(stderr, "Failed to issue method call: %s\n", error.message);
+                sd_bus_unref(bus);
+                return r;
+        }
+
+        r = sd_bus_message_enter_container(result, SD_BUS_TYPE_ARRAY, "(soss)");
+        if (r < 0) {
+                fprintf(stderr, "Failed to open result array: %s\n", strerror(-r));
+                sd_bus_unref(bus);
+                sd_bus_message_unref(result);
+                sd_bus_error_free(&error);
+                return r;
+        }
+        while (sd_bus_message_at_end(result, false) == 0) {
+                const char *name = NULL;
+
+                r = sd_bus_message_read(result, "(soss)", &name, NULL, NULL, NULL);
+                if (r < 0) {
+                        fprintf(stderr, "Failed to read node information: %s\n", strerror(-r));
+                        sd_bus_unref(bus);
+                        sd_bus_message_unref(result);
+                        sd_bus_error_free(&error);
+                        return r;
+                }
+
+                char *node_name = strdup(name);
```

Comment:
You need to free the memory allocated with`strdup`:
```suggestion
                _cleanup_free_ char *node_name = strdup(name);
```